File: 0374-guess-number-higher-or-lower/0374-guess-number-higher-or-lower.kt
/** 
 * The API guess is defined in the parent class.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * fun guess(num:Int):Int {}
 */

class Solution:GuessGame() {
    override fun guessNumber(n:Int):Int {
        var low = 1
        var high = n
        
        while (low <= high) {
            val myGuess = (high - low) / 2 + low
            when (guess(myGuess)) {
                -1 -> high = myGuess - 1
                1 -> low = myGuess + 1
                0 -> return myGuess
            }
        }
        return low
    }
}

File: 0374-guess-number-higher-or-lower/0374-guess-number-higher-or-lower.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        lo, hi = 0, n+1
        while hi-lo > 1:
            mid = (lo + hi)//2
            match (guess(mid)):
                case 1: lo = mid
                case -1: hi = mid
                case _: return mid
        return -1
        